docs/specification/draft/basic/transports.mdx

```diff
@@ -287,6 +288,164 @@ protocol version 2024-11-05) as follows:
+- Upon reconnection, the client calls initialize with a `sessionId` and the server **MUST** close any still open WebSocket connection associated with that session ID with close code `1000` (Normal Closure) and reason "MCP session reconnected on another connection"
+- Close frames might be ignored and there is delay in effectualization, so the server **MUST** also ignore any messages that arrive for the MCP session on any connection but the active one
+- Server-initiated notifications **MUST** be sent over the active connection
+- Clients **SHOULD** keep track of unacknowledged messages and resend them if necessary after reconnection
+
+### WebSocket Transport Lifecycle
+
+#### Initial Connection and Session Establishment
+
+**Client Behavior:**
+- Establish WebSocket connection to server
+- Send MCP `initialize` without `sessionId`
+- Wait for `initialize` response and extract `sessionId`
+- Store `sessionId` for reconnection
+- Send `notifications/initialized` with `sessionId` [TODO: SEP-1364 envisions making this optional. Adjust once finalized]
+
+**Server Behavior:**
+- Accept WebSocket connection
+- Generate and include `sessionId` when responding to `initialize` request
+
+#### Normal Operation
+
+During normal operation:
+- Client **MUST** include the `sessionId` in all messages except the first `initialize`
+- Server **MUST** close the WebSocket connection with close code `1008` (Policy Violation) and reason "Missing sessionId" for requests without `sessionId` (other than `initialize`)
+- Server maintains session state and processes requests normally
+- Server may send notifications to the client at any time
+- Both client and server should handle JSON-RPC requests and responses according to MCP specification
+
+#### Disconnection Handling
+
+WebSocket connection interruption might occur at any time (e.g., due to network conditions). Therefore:
+- WebSocket connection interruption **MUST NOT** be interpreted as the client cancelling its request
+- To cancel, the client **SHOULD** explicitly send an MCP `CancelledNotification`
+- Similarly, disconnection **MUST NOT** be interpreted as terminating the MCP session
+- Server **MUST** preserve session state for a reasonable timeout period to allow reconnection
+
+#### Reconnection and Session Recovery
+
+Upon reconnection:
+- Client **MUST** include the `sessionId` from the previous session in an `initialize` call after reconnection
+- Server **MUST** recognize the existing `sessionId` and associate this new connection with previously established session state
+- Server **MUST** close any previous WebSocket connection associated with the same MCP `sessionId` with close code `1000` (Normal Closure) and reason "MCP session reconnected on another connection"
+- Normal operation can resume immediately after MCP session recognition
```

Comment:
If the session ID isn't recognized, what happens? Does the server issue a new session ID in the initialize response? That would be a nice way for servers that aren't reconnection-aware to "just work" if they don't look at the presented session ID.

Reply:
I have removed any content that duplicates the behavior defined in SEP-1364. However, I believe session IDs will continue to be optional with SEP-1364, and it may not specify that servers can require them. I want sessions to be required for this WebSocket transport for various reasons. WebSockets are only needed for bidirectional communication, which requires sessions. If you desire a sessionless approach, you can use HTTP POST. I also prefer "one good way to do things". So, we'll have to adjust once SEP-1364 is finalized.